# 아이템 70) 복구할 수 있는 상황에는 검사 예외를, 프로그래밍 오류에는 런타임 예외를 사용하라

### Object - Throwable - Error / Exception 관계

![](https://www.notion.so/image/https%3A%2F%2Fprod-files-secure.s3.us-west-2.amazonaws.com%2F8c65f92d-6f83-471d-8b8b-2101f652bce7%2Fd7318c18-9414-42b0-b66b-72071d147bba%2FUntitled.png?table=block&id=7565bab0-8f73-4967-9829-814e41fe6953&spaceId=8c65f92d-6f83-471d-8b8b-2101f652bce7&width=2000&userId=edab0db8-cc3f-4181-b4ba-153fc87185e3&cache=v2)

![](https://www.notion.so/image/https%3A%2F%2Fprod-files-secure.s3.us-west-2.amazonaws.com%2F8c65f92d-6f83-471d-8b8b-2101f652bce7%2F8d608310-e4de-45a2-b5af-08a3327cd8ef%2FUntitled.png?table=block&id=12d77e12-0a8b-4b69-8cb6-21a8eb1dad1d&spaceId=8c65f92d-6f83-471d-8b8b-2101f652bce7&width=2000&userId=edab0db8-cc3f-4181-b4ba-153fc87185e3&cache=v2)

- `Error` : 시스템 레벨의 심각한 수준의 에러
    - 개발자가 미리 예측할 수도 없고 처리할 수 있는 방법도 없음
    - 애플리케이션 단에서는 Error에 대한 처리를 신경 쓰지 않아도 됨
        - OutOfMemoryError이나 ThreadDeath 같은 에러는 try catch으로 잡아도 할 수 있는 것이 없기 때문
- `Exception` : 개발자가 로직을 추가하여 처리 가능
    - `Checked Exception` : 애플리케이션이 예상하고 복구해야하는 예외적인 조건
    - `Unchecked Exception` : 애플리케이션 내부적으로 예외적인 상황이므로 일반적으로 이 예외를 예상하거나 복구할 수 없음
    
    - `RuntimeException` ; Checked / Unchecked Exception 을 구분하는 기준
    - 컴파일 단계에서 확인할 수 없고 (Unchecked Exception) 실행 과정 중에 발견되기 때문에 (Runtime Exception)
    
    ![](https://www.notion.so/image/https%3A%2F%2Fprod-files-secure.s3.us-west-2.amazonaws.com%2F8c65f92d-6f83-471d-8b8b-2101f652bce7%2Fc7f4ec26-d17e-4cbb-867f-f82773de42b2%2FUntitled.png?table=block&id=52408b37-2125-4399-9c6e-9cd5b97dfb32&spaceId=8c65f92d-6f83-471d-8b8b-2101f652bce7&width=2000&userId=edab0db8-cc3f-4181-b4ba-153fc87185e3&cache=v2)

    - Checked Exception 발생 시 rollback 되지 않고 트랜잭션 commit 까지 완료됨

### ****왜 Checked Exception은 Rollback되지 않는 것일까?****

- 기본적으로 Checked Exception는 복구가 가능하다는 메커니즘을 가지고 있음
    
    ex) 특정 이미지 파일을 찾아서 전송해주는 함수에서 이미지를 찾지 못했을 경우 기본 이미지를 전송 <복구 전략>
    
    ```java
    public void sendFile(String fileName){
        File file;
        try {
            file = FileFindService.find(fileName);
        } catch (FileNotFoundException e){ // FileNotFoundException은 IOException으로 checked exception
            // 파일을 못찾았으니 기본 파일을 찾아서 전송
            file = FileFindService.find("default.png");
        }
        send(file);
    }
    ```
    
    - 기본적으로 복구가 가능하니 네가 복구를 작업을 진행했을 수 있으니까 Rollback은 진행하지 않을게라는 의미가 있다고 생각 (누군가의 주관적 생각)
    - 사실 이런식의 예외는 복구가 아니라, 일반 코드의 흐름으로 제어해야 함
    
    ```java
    public void sendFile(String fileName){
        if(FileFindService.existed(filename)){
            // 파일이 있는 경우 해당 파일을 찾아서 전송
            send(FileFindService.find(fileName));    
        }else{
            // 파일이 있는 없는 경우 기본 이미지 전송
            send(FileFindService.find("default.png"));    
        }
    }
    ```
    
- 하지만 일반적으로 Checked Exception 예외가 발생했을 경우 복구 전략을 갖고 그것을 복구할 수 있는 경우는 그렇게 많지 않음
    
    ex) 유니크해야 하는 이메일 값이 중복돼서 SQLException이 발생하는 경우 
    
    - 유저가 압력을 가했던 이메일 + 난수를 입력해서 insert 시키면 가능은 하겠지만, RuntimeException을 발생시키고 입력을 다시 유도하는 것이 현실적
    - **여기서 중요한 것은 해당 Exception을 발생시킬 때 명확하게 어떤 예외가 발생해서 Exception이 발생했는지 정보를 전달해주는 것**
    - 이 예시의 경우, **DuplicateEmailException (Unchecked Exception)을 발생 시키는 것이 바람직함**
- 즉, Checked Exception을 만나면 더 구체적인 **Unchecked Exception을 발생시켜 정확한 정보를 전달하고 로직의 흐름을 끊어야 함**
    - 해당 예러가 왜 발생했는지에 대해서 에러 메시지 뿐만이 아니라 더욱 구체적인 정보를 전달해주는것이 좋음
    
    ```java
    public <T> T readValue(String json, Class<T> clazz) {
        try {
          return objectMapper.readValue(json, clazz);
        } catch (IOException e) {
          throw new JsonDeserializeFailed(e.getMessage());
        }
      }
    ```
    
    - 실제 예시) JPA에 구현체를 가져다 사용하더라도 Checked Exception을 직접 처리하지 않고 있는 이유도 다 적절한 RuntimeException으로 예외를 던져주고 있기 때문

### 책에서 나오는 단어들을 간단히 정리하자면,

1. `검사 (Checked)`  ; 복구할 수 있는 상황일 때 사용
    - 검사 예외 (Checked Exception)
2. `비검사 (Unchecked throwable)` ; 프로그램에서 잡을 필요 없거나 혹은 통상적으로는 잡지 말아야 함 (복구가 불가능하거나, 더 실행해봐야 실이 많음)
    - 런타임 예외 (Unchecked Exception == Runtime Exception)
    - 에러 (Error)
3. `Error, Exception, RuntimeException 을 상속하지 않는 throwable` ; 사용하지 말자
    - 암묵적으로 검사 예외처럼 다룸

### **검사 예외 (Checked Exception)**

- 호출하는 쪽에서 복구하리라 여겨지는 상황이라면 이것을 사용
- checked exception 은 호출자가 그 예외를 **catch로 잡아 처리하거나, 위로 전파하도록 강제**한다.
    - 메서드 선언에 포함된 checked exception 각각은 그 **메서드를 호출했을 때 발생할 수 있는 유력한 결과임을 클라이언트에게 알려주는 것**
- 검사 예외는 **복구에 필요한 정보를 제공하는 메서드도 제공**하자.
    
    ```java
    class CustomException extends Exception {
        private String additionalInfo;
    
        // 생성자를 통해 예외 메시지와 추가 정보를 전달받음
        public CustomException(String message, String additionalInfo) {
            super(message);
            this.additionalInfo = additionalInfo;
        }
    
        // 추가 정보를 얻기 위한 메서드
        public String getAdditionalInfo() {
            return additionalInfo;
        }
    }
    
    public class ExceptionExample {
        public static void main(String[] args) {
            try {
                // 예외 발생 시 추가 정보를 함께 제공
                throwCustomException();
            } catch (CustomException e) {
                // 검사 예외를 처리하면서 추가 정보를 활용
                System.out.println("Caught CustomException: " + e.getMessage());
                System.out.println("Additional Info: " + e.getAdditionalInfo());
            }
        }
    
        // 예외 발생 메서드
        public static void throwCustomException() throws CustomException {
            String additionalInfo = "This is additional information.";
            throw new CustomException("Custom exception occurred", additionalInfo);
        }
    }
    ```
    

### **런타임 예외 (Unchecked Exception)**

- **프로그래밍 오류를 나타낼 때는 런타임 예외를 사용하자.**
- 런타임 예외의 대부분은 전제조건을 만족하지 못했을 때 발생한다. 
(ex. 배열 인덱스 크기. API 명세에 기록된 제약을 지키지 못했다는 것)
    
    ```java
    public class Car {
    	private String name;
    
    	public Car(String name) {
    		if(name.length() > 5){
    			throw new IllegalArgumentException("이름은 5자를 넘을 수 없습니다.");
    		}
    		this.name = name;
    	}
    }
    ```
    

### 에러

- 에러는 보통 더 이상 수행할 수 없는 상황을 나타낸다. ex) JVM 의 자원부족, 불변식 깨짐 등
- Error 클래스의 의도는 명확하므로 Error 클래스를 상속해 하위 클래스를 만드는 일은 자제하자.
즉, **비검사 예외의 경우는 RuntimeException만 상속해서 만들라**
- Error는 상속하지도 말고, throw문도 던지지 말자 (AssertionError는 예외)

### 결론

- 복구할 수 있다면, 검사 예외 (Checked Exception)
- 프로그래밍 오류라면 (복구가 불가능하다면) 비검사 예외 (Unchecked Exception)
- 확실하지 않다면 비검사 예외 (Unchecked Exception)
- 검사 예외도 아니고 런타임 예외도 아닌 throwable은 정의하지도 말자. (Exception, RuntimeException, Error 를 상속하지 않는 throwable)
- 재미있는 글 : [https://velog.io/@eastperson/Java의-Checked-Exception은-실수다-83omm70j](https://velog.io/@eastperson/Java%EC%9D%98-Checked-Exception%EC%9D%80-%EC%8B%A4%EC%88%98%EB%8B%A4-83omm70j)